### DISCUSSION TITLE
Using a local Modrinth .mrpack isn't working ("Failed to read modpack index: Archive is not a ZIP archive")

### DISCUSSION THREAD

**Discussion by templeofshadow** [itzg/docker-minecraft-server#3210] [Q&A] [answered]
Hello all. I'm new to using Docker (and running servers in general) but I recently got the opportunity to spin up a server on my old machine, and this looked like the perfect tool to do it.

After several hours of trial and error, I finally decided to ask here and see if I was doing anything wrong that I hadn't noticed.

Here is my docker-compose.yaml (with some redaction):
```
services:
  mc:
    image: itzg/minecraft-server
    tty: true
    stdin_open: true
    ports:
      - "25565:25565"
    environment:
      MODPACK_PLATFORM: "MODRINTH"
      MODRINTH_MODPACK: "/modpacks/1_21_1-personal-neoforge-modpack-0_0_2.mrpack"
      EULA: "TRUE"
      MEMORY: "8G"
      RCON_PASSWORD: "[REDACTED]"
      MOTD: "temple's server"
      OPS: "TempleOfShadows"
      DEBUG: true
    volumes:
      - "./data:/data"
      - "./modpacks:/modpacks:ro"
```

And here is the log I am getting from trying to run it:

```
mc-1  | + export HOME=/data
mc-1  | + HOME=/data
mc-1  | ++ id -u
mc-1  | ++ id -g
mc-1  | ++ ls -lnd /data
mc-1  | + log 'Running as uid=1000 gid=1000 with /data as '\''drwxr-xr-x 2 1000 1000 4096 Dec 26 09:04 /data'\'''
mc-1  | + local oldState
mc-1  | ++ shopt -po xtrace
mc-1  | + oldState='set -o xtrace'
mc-1  | + shopt -u -o xtrace
mc-1  | [init] 2024-12-26 09:07:48+00:00 Running as uid=1000 gid=1000 with /data as 'drwxr-xr-x 2 1000 1000 4096 Dec 26 09:04 /data'
mc-1  | + '[' '!' -e /data/eula.txt ']'
mc-1  | + isTrue false
mc-1  | + case "${1,,}" in
mc-1  | + return 1
mc-1  | + isTrue true
mc-1  | + case "${1,,}" in
mc-1  | + return 0
mc-1  | + set +x
mc-1  | + isTrue false
mc-1  | + case "${1,,}" in
mc-1  | + return 1
mc-1  | + isTrue false
mc-1  | + case "${1,,}" in
mc-1  | + return 1
mc-1  | + [[ -n '' ]]
mc-1  | + fixJavaPath
mc-1  | + which java
mc-1  | + cd /data
mc-1  | + export DECLARED_TYPE=VANILLA
mc-1  | + DECLARED_TYPE=VANILLA
mc-1  | + export DECLARED_VERSION=LATEST
mc-1  | + DECLARED_VERSION=LATEST
mc-1  | + isTrue false
mc-1  | + case "${1,,}" in
mc-1  | + return 1
mc-1  | + isTrue false
mc-1  | + case "${1,,}" in
mc-1  | + return 1
mc-1  | + [[ -n '' ]]
mc-1  | + [[ -n '' ]]
mc-1  | + [[ -n '' ]]
mc-1  | + [[ -n '' ]]
mc-1  | + [[ -n '' ]]
mc-1  | + : MODRINTH
mc-1  | + [[ -n MODRINTH ]]
mc-1  | + [[ -n VANILLA ]]
mc-1  | + [[ VANILLA != VANILLA ]]
mc-1  | + case "${TYPE^^}" in
mc-1  | + [[ -n MODRINTH ]]
mc-1  | + case "${MODPACK_PLATFORM^^}" in
mc-1  | + exec /start-deployModrinth
mc-1  | + ensureRemoveAllModsOff MODPACK_PLATFORM=MODRINTH
mc-1  | + reason=MODPACK_PLATFORM=MODRINTH
mc-1  | + isTrue false
mc-1  | + case "${1,,}" in
mc-1  | + return 1
mc-1  | + args=(--results-file="$resultsFile" --project="${MODRINTH_MODPACK}" --output-directory=/data)
mc-1  | + [[ -n '' ]]
mc-1  | + case "${VERSION^^}" in
mc-1  | + : release
mc-1  | + setArg --loader MODRINTH_LOADER
mc-1  | + arg=--loader
mc-1  | + var=MODRINTH_LOADER
mc-1  | + [[ -n '' ]]
mc-1  | + setArg --version MODRINTH_VERSION
mc-1  | + arg=--version
mc-1  | + var=MODRINTH_VERSION
mc-1  | + [[ -n '' ]]
mc-1  | + setArg --default-version-type MODRINTH_DEFAULT_VERSION_TYPE
mc-1  | + arg=--default-version-type
mc-1  | + var=MODRINTH_DEFAULT_VERSION_TYPE
mc-1  | + [[ -n release ]]
mc-1  | + args+=("${arg}=${!var}")
mc-1  | + setArg --exclude-files MODRINTH_EXCLUDE_FILES
mc-1  | + arg=--exclude-files
mc-1  | + var=MODRINTH_EXCLUDE_FILES
mc-1  | + [[ -n '' ]]
mc-1  | + setArg --force-include-files MODRINTH_FORCE_INCLUDE_FILES
mc-1  | + arg=--force-include-files
mc-1  | + var=MODRINTH_FORCE_INCLUDE_FILES
mc-1  | + [[ -n '' ]]
mc-1  | + setArg --overrides-exclusions MODRINTH_OVERRIDES_EXCLUSIONS
mc-1  | + arg=--overrides-exclusions
mc-1  | + var=MODRINTH_OVERRIDES_EXCLUSIONS
mc-1  | + [[ -n '' ]]
mc-1  | + setArg --default-exclude-includes MODRINTH_DEFAULT_EXCLUDE_INCLUDES
mc-1  | + arg=--default-exclude-includes
mc-1  | + var=MODRINTH_DEFAULT_EXCLUDE_INCLUDES
mc-1  | + [[ -n /image/modrinth-exclude-include.json ]]
mc-1  | + args+=("${arg}=${!var}")
mc-1  | + mc-image-helper install-modrinth-modpack --results-file=/data/.install-modrinth.env --project=/modpacks/1_21_1-personal-neoforge-modpack-0_0_2.mrpack --output-directory=/data --default-version-type=release --default-exclude-includes=/image/modrinth-exclude-include.json
mc-1  | [mc-image-helper] 09:07:49.550 ERROR : 'install-modrinth-modpack' command failed. Version is 1.40.6
mc-1  | me.itzg.helpers.errors.GenericException: Failed to read modpack index: Archive is not a ZIP archive
mc-1  |         at me.itzg.helpers.modrinth.ModrinthPackInstaller.processModpack(ModrinthPackInstaller.java:92)
mc-1  |         at me.itzg.helpers.modrinth.InstallModrinthModpackCommand.lambda$call$4(InstallModrinthModpackCommand.java:152)
mc-1  |         at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:132)
mc-1  |         at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:79)
mc-1  |         at reactor.core.publisher.Operators$ScalarSubscription.request(Operators.java:2571)
mc-1  |         at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.set(Operators.java:2367)
mc-1  |         at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onSubscribe(FluxOnErrorResume.java:74)
mc-1  |         at reactor.core.publisher.MonoJust.subscribe(MonoJust.java:55)
mc-1  |         at reactor.core.publisher.Mono.subscribe(Mono.java:4576)
mc-1  |         at reactor.core.publisher.Mono.block(Mono.java:1778)
mc-1  |         at me.itzg.helpers.modrinth.InstallModrinthModpackCommand.call(InstallModrinthModpackCommand.java:169)
mc-1  |         at me.itzg.helpers.modrinth.InstallModrinthModpackCommand.call(InstallModrinthModpackCommand.java:29)
mc-1  |         at picocli.CommandLine.executeUserObject(CommandLine.java:2045)
mc-1  |         at picocli.CommandLine.access$1500(CommandLine.java:148)
mc-1  |         at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2465)
mc-1  |         at picocli.CommandLine$RunLast.handle(CommandLine.java:2457)
mc-1  |         at picocli.CommandLine$RunLast.handle(CommandLine.java:2419)
mc-1  |         at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2277)
mc-1  |         at picocli.CommandLine$RunLast.execute(CommandLine.java:2421)
mc-1  |         at picocli.CommandLine.execute(CommandLine.java:2174)
mc-1  |         at me.itzg.helpers.McImageHelper.main(McImageHelper.java:161)
mc-1  |         Suppressed: java.lang.Exception: #block terminated with an error
mc-1  |                 at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:104)
mc-1  |                 at reactor.core.publisher.Mono.block(Mono.java:1779)
mc-1  |                 ... 11 common frames omitted
mc-1  | Caused by: java.util.zip.ZipException: Archive is not a ZIP archive
mc-1  |         at org.apache.commons.compress.archivers.zip.ZipFile.positionAtEndOfCentralDirectoryRecord(ZipFile.java:562)
mc-1  |         at org.apache.commons.compress.archivers.zip.ZipFile.openZipChannel(ZipFile.java:504)
mc-1  |         at org.apache.commons.compress.archivers.zip.ZipFile.access$000(ZipFile.java:89)
mc-1  |         at org.apache.commons.compress.archivers.zip.ZipFile$Builder.get(ZipFile.java:159)
mc-1  |         at me.itzg.helpers.files.IoStreams.readFileFromZip(IoStreams.java:25)
mc-1  |         at me.itzg.helpers.modrinth.ModrinthPackInstaller.processModpack(ModrinthPackInstaller.java:87)
mc-1  |         ... 20 common frames omitted
mc-1  | + logError 'Failed to installModrinth modpack'
mc-1  | + isDebugging
mc-1  | + isTrue true
mc-1  | + case "${1,,}" in
mc-1  | + return 0
mc-1  | + set +x
mc-1  | [init] 2024-12-26 09:07:49+00:00 [ERROR] Failed to installModrinth modpack
mc-1  | + exit 1
```

You can find the .mrpack file [here](https://github.com/templeofshadow/mc-server-fileshare/blob/186d23c2a90bb7cc422a2b94d82bd4663f836acf/modrinth%20modpack%20backup/1.21.1%20personal%20neoforge%20modpack%200.0.2.mrpack) (I tried pulling from this online repo for ease of updating but that runs into the same issue).

Am I doing anything obviously wrong? Am I misunderstanding how this works? I scoured the documentation and it seems to me that I'm doing everything correctly...

In any case, I look forward to any help you all might be able to provide. Thanks.

EDIT: I can't believe I forgot to mention this; Pulling from the Modrinth website works just fine, and so does using a vanilla install. I just run into this problem using my own file, so I personally suspect that I have done something wrong there.

**Comment by itzg** [comment]
Did you make sure to download the raw file (vs the github page)

https://github.com/templeofshadow/mc-server-fileshare/raw/186d23c2a90bb7cc422a2b94d82bd4663f836acf/modrinth%20modpack%20backup/1.21.1%20personal%20neoforge%20modpack%200.0.2.mrpack

And is that mrpack file a zip file?

**Reply by templeofshadow** [reply]
> Did you make sure to download the raw file (vs the github page)

Yes, when I tried to use the link instead of a local file, I did that via the raw link. It did the same.

Everything that I posted in the initial discussion post is about using a local .mrpack file, though, so I doubt that there is an issue there personally.

> And is that mrpack file a zip file?

I'm sorry, but what exactly do you mean by this? 

To iterate through my process in getting the file, I had a modpack instance in the Modrinth app, exported it, and then placed that file in the GitHub repo. I honestly have no idea about what exactly is in the mrpack file, I just know it worked how I expected when I tried importing it back into the Modrinth app. So, I have no idea if the mrpack file is a zip file. All I know is that the file extension is .mrpack.

I did try putting that mrpack file into a zip archive and using that instead, but it didn't seem to work.

EDIT: 

Okay, so I reviewed the Modrinth modpack format spec, specifically the storage section ([this](https://support.modrinth.com/en/articles/8802351-modrinth-modpack-format-mrpack#h_fa7dccdcdd)), and seeing that the .mrpack format is basically just a ZIP archive, I checked my file by opening it with 7-zip and checking it against a known-good modpack (an example modpack downloaded directly from Modrinth, that I used when testing if the online functionality worked), and they looked the same.

**Reply by templeofshadow** [reply]
Hi @itzg I just wanted to follow up on this, noting that the answers to your original response are "Yes (although not applicable in this case)" and "Yes, to the best of my knowledge", is there anything else that I might be able to look into in order to fix this issue?

I haven't had the time to dive in-depth into this so I'd appreciate it if you could point me in a direction (or better yet, the solution outright). In any case, thank you for putting in the time to respond originally.

**Reply by itzg** [reply]
I also haven't had a chance to look further. All I know is that it says "Archive is not a ZIP archive" but that's strange if you indeed exported the mrpack file from the Modrinth client. 

mrpack files are zip files

https://support.modrinth.com/en/articles/8802351-modrinth-modpack-format-mrpack

**Reply by templeofshadow** [reply]
All right, thanks for the quick reply. I'll hopefully get a moment soon to look into this myself and figure it out.

If/when I figure out what's going on, I'll post a reply here.

**Reply by templeofshadow** [reply, marked as answer]
I now feel quite stupid.

For some unknown reason, when I downloaded the .mrpack file to my server, it deleted all of the data it contained, leaving a 0kb .mrpack file of the same name. I never noticed this, because I was using ssh from another machine and the terminal doesn't tell me the sizes of files when using ls (at least, not natively). 

In any case, since it was obviously an issue with the file, I finally wizened up and used rsync to pull the file to my main machine and check it against the original, and promptly noticed the issue. I then directly copied a freshly exported file to the server (again using rsync) and then tried it again. It now works as expected (that is to say, it errors out, but that's because I need to refine the modpack configs for the server instance, not due to any issues with the docker image).
